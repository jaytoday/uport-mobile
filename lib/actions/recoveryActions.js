// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import {
  SHOW_RECOVERY_SEED,
  SAVE_RECOVERY_SEED,
  HIDE_RECOVERY_SEED,
  SELECT_RECOVERY_SEED_WORD,
  START_SEED_RECOVERY,
  RECOVERY_SUCCESS
} from 'uPortMobile/lib/constants/RecoveryActionTypes'
// import { SOCIAL_RECOVERY_POP_ROUTE, SOCIAL_RECOVERY_PUSH_ROUTE, SOCIAL_RECOVERY_RESET_ROUTE } from '../constants/SocialRecoveryActionTypes'

export function showRecoverySeed () {
  return {
    type: SHOW_RECOVERY_SEED
  }
}

export function saveRecoverySeed (seed) {
  return {
    type: SAVE_RECOVERY_SEED,
    seed
  }
}

export function hideRecoverySeed () {
  return {
    type: HIDE_RECOVERY_SEED
  }
}

export function selectRecoveryWord (index) {
  return {
    type: SELECT_RECOVERY_SEED_WORD,
    index
  }
}

export function startSeedRecovery (phrase, componentId) {
  return {
    type: START_SEED_RECOVERY,
    phrase,
    componentId
  }
}

export function recoverySuccess () {
  return {
    type: RECOVERY_SUCCESS
  }
}
