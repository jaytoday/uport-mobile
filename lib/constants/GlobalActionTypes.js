// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
export const RESET_DEVICE = 'RESET_DEVICE'
export const RELOAD_DB = 'RELOAD_DB'
export const LOADED_DB = 'LOADED_DB'
export const PERSIST_DB = 'PERSIST_DB'
export const FORCE_PERSIST_DB = 'FORCE_PERSIST_DB'
