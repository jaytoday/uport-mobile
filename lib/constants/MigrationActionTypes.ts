import MigrationScreen from '../components/Migrations/MigrationScreen'

//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//

export const ADD_MIGRATION_TARGET = 'ADD_MIGRATION_TARGET'
export const RUN_MIGRATIONS = 'RUN_MIGRATIONS'
export const STARTED_MIGRATION_STEP = 'STARTED_MIGRATION_STEP'
export const COMPLETED_MIGRATION_STEP = 'COMPLETED_MIGRATION_STEP'
export const FAILED_MIGRATION_STEP = 'FAILED_MIGRATION_STEP'

import SCREENS from 'uPortMobile/lib/screens/Screens'

export enum MigrationTarget {
  Legacy = 'Legacy',
  MissingKeys = 'MissingKeys',
  RecoverSeed = 'RecoverSeed',
}

export enum MigrationStep {
  MigrateLegacy = 'MigrateLegacy',
  EnterSeed = 'EnterSeed',
}

export enum MigrationStatus {
  NotStarted,
  Started,
  Completed,
  Error,
}

export interface Recipes {
  [index: string]: MigrationStep[]
}

export const targetRecipes: Recipes = {
  Legacy: [MigrationStep.MigrateLegacy],
  RecoverSeed: [MigrationStep.EnterSeed],
}

export interface MigrationScreens {
  [index: string]: string
}

export const migrationScreens: MigrationScreens = {
  RecoverSeed: SCREENS.RECOVERY.RestoreSeedInstructions,
}

export interface TargetAction {
  type: 'ADD_MIGRATION_TARGET' | 'RUN_MIGRATIONS'
  target: MigrationTarget
  _backup?: boolean
}

export interface StepAction {
  type: 'STARTED_MIGRATION_STEP' | 'COMPLETED_MIGRATION_STEP' | 'FAILED_MIGRATION_STEP'
  step: MigrationStep
  _backup?: boolean
}
