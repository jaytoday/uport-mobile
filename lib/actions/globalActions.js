// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { RESET_DEVICE, RELOAD_DB, PERSIST_DB, FORCE_PERSIST_DB, LOADED_DB } from '../constants/GlobalActionTypes'

export function resetDevice () {
  return { type: RESET_DEVICE }
}

export function reload (db) {
  return {
    type: RELOAD_DB,
    db
  }
}

// trigger persistence of a certain reducer (the key in the global reducer. eg 'uport')
export function persist (reducer) {
  return {
    type: PERSIST_DB,
    reducer
  }
}

// trigger persistence of a certain reducer (the key in the global reducer. eg 'uport')
export function forcePersistDB () {
  return {
    type: FORCE_PERSIST_DB
  }
}

export function loadedDB () {
  return { type: LOADED_DB }
}
