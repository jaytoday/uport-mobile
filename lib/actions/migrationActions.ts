//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//

import { ADD_MIGRATION_TARGET, STARTED_MIGRATION_STEP, COMPLETED_MIGRATION_STEP, FAILED_MIGRATION_STEP, RUN_MIGRATIONS, MigrationStep, MigrationTarget, TargetAction, StepAction } from 'uPortMobile/lib/constants/MigrationActionTypes'

const _backup = true

export function addMigrationTarget(target: MigrationTarget) : TargetAction {
  return {
    type: ADD_MIGRATION_TARGET,
    target,
    _backup
  }
}

export function runMigrations(target: MigrationTarget) : TargetAction {
  return {
    type: RUN_MIGRATIONS,
    target
  }
}

export function startedMigrationStep(step: MigrationStep) : StepAction {
  return {
    type: STARTED_MIGRATION_STEP,
    step,
    _backup
  }
}

export function completedMigrationStep(step: MigrationStep) : StepAction {
  return {
    type: COMPLETED_MIGRATION_STEP,
    step,
    _backup
  }
}

export function failedMigrationStep(step: MigrationStep) : StepAction {
  return {
    type: FAILED_MIGRATION_STEP,
    step,
    _backup
  }
}