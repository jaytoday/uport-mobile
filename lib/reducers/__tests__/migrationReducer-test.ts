// This file is part of the uPort Mobile App.

// The uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// The uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with The uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.

// (C) Copyright 2016-2018 ConsenSys AG

import { addMigrationTarget, startedMigrationStep, completedMigrationStep, failedMigrationStep } from 'uPortMobile/lib/actions/migrationActions'
import { MigrationStep, MigrationTarget, MigrationStatus } from 'uPortMobile/lib/constants/MigrationActionTypes'
import reducer, { MigrationState } from '../migrationsReducer'


const initialState : MigrationState = {
  targets: [],
  steps: {}
}

const populated : MigrationState = {
  targets: [MigrationTarget.Legacy],
  steps: {}
}
populated.steps[MigrationStep[MigrationStep.MigrateLegacy]] = MigrationStatus.Started

describe('addMigrationTarget', () => {
  it('adds a target', () => {
    expect(reducer(initialState, addMigrationTarget(MigrationTarget.Legacy))).toMatchSnapshot()
  })

  it('does not add an existing target', () => {
    expect(reducer(populated, addMigrationTarget(MigrationTarget.Legacy))).toEqual(populated)
  })
})

describe('startedMigrationStep', () => {
  it('starts a step', () => {
    expect(reducer(initialState, startedMigrationStep(MigrationStep.MigrateLegacy))).toMatchSnapshot()
  })

  it('does not a step if already started', () => {
    expect(reducer(populated, startedMigrationStep(MigrationStep.MigrateLegacy))).toEqual(populated)
  })

  it('should start an additional step', () => {
    expect(reducer(populated, startedMigrationStep(MigrationStep.MigrateLegacy))).toMatchSnapshot()
  })
})

describe('completedMigrationStep', () => {
  it('completes a step', () => {
    expect(reducer(initialState, completedMigrationStep(MigrationStep.MigrateLegacy))).toMatchSnapshot()
  })

  it('completes a step if already started', () => {
    expect(reducer(populated, completedMigrationStep(MigrationStep.MigrateLegacy)).steps[MigrationStep.MigrateLegacy]).toEqual(MigrationStatus.Completed)
  })

  it('should complete an additional step', () => {
    expect(reducer(populated, completedMigrationStep(MigrationStep.MigrateLegacy))).toMatchSnapshot()
  })
})

describe('failedMigrationStep', () => {
  it('fails a step', () => {
    expect(reducer(initialState, failedMigrationStep(MigrationStep.MigrateLegacy))).toMatchSnapshot()
  })

  it('fails a step if already started', () => {
    expect(reducer(populated, failedMigrationStep(MigrationStep.MigrateLegacy)).steps[MigrationStep.MigrateLegacy]).toEqual(MigrationStatus.Error)
  })

  it('should fail an additional step', () => {
    expect(reducer(populated, failedMigrationStep(MigrationStep.MigrateLegacy))).toMatchSnapshot()
  })
})
