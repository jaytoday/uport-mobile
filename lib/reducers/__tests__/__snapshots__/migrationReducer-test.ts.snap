// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addMigrationTarget adds a target 1`] = `
Object {
  "steps": Object {},
  "targets": Array [
    "Legacy",
  ],
}
`;

exports[`completedMigrationStep completes a step 1`] = `
Object {
  "steps": Object {
    "MigrateLegacy": 2,
  },
  "targets": Array [],
}
`;

exports[`completedMigrationStep should complete an additional step 1`] = `
Object {
  "steps": Object {
    "MigrateLegacy": 2,
  },
  "targets": Array [
    "Legacy",
  ],
}
`;

exports[`failedMigrationStep fails a step 1`] = `
Object {
  "steps": Object {
    "MigrateLegacy": 3,
  },
  "targets": Array [],
}
`;

exports[`failedMigrationStep should fail an additional step 1`] = `
Object {
  "steps": Object {
    "MigrateLegacy": 3,
  },
  "targets": Array [
    "Legacy",
  ],
}
`;

exports[`startedMigrationStep should start an additional step 1`] = `
Object {
  "steps": Object {
    "MigrateLegacy": 1,
  },
  "targets": Array [
    "Legacy",
  ],
}
`;

exports[`startedMigrationStep starts a step 1`] = `
Object {
  "steps": Object {
    "MigrateLegacy": 1,
  },
  "targets": Array [],
}
`;
